core/src/main/java/io/questdb/cairo/ColumnVersionPurgeJob.java

```diff
@@ -0,0 +1,349 @@
+            boolean useful = processInQueue();
+            boolean cleanupUseful = cleanup();
+            if (cleanupUseful) {
+                LOG.debug().$("cleaned column version, outstanding tasks: ").$(houseKeepingRunQueue.size()).$();
+            }
+            inErrorCount = 0;
+            return cleanupUseful || useful;
+        } catch (Throwable th) {
+            LOG.error().$("failed to clean up column versions").$(th).$();
+            inErrorCount++;
+            if (inErrorCount == MAX_ERRORS) {
+                if (houseKeepingRunQueue.size() > 0) {
+                    LOG.error().$("clean up column versions reached maximum error count and will be recycled. Some column version may be left behind.").$(th).$();
+                    houseKeepingRunQueue.clear();
+                    inErrorCount = 0;
+                } else {
+                    LOG.error().$("clean up column versions reached maximum error count and will be DISABLED. Restart QuestDB to re-enable the job.").$(th).$();
+                    close();
+                }
+            }
+            return false;
+        }
+    }
+
+    private void saveToStorage(ColumnVersionPurgeTaskRun cleanTask) {
+        if (writer != null) {
+            try {
+                LongList updatedColumnVersions = cleanTask.getUpdatedColumnVersions();
+                for (int i = 0, n = updatedColumnVersions.size(); i < n; i += ColumnVersionPurgeTask.BLOCK_SIZE) {
+                    TableWriter.Row row = writer.newRow(cleanTask.timestamp);
+                    row.putSym(TABLE_NAME_COLUMN, cleanTask.getTableName());
+                    row.putSym(COLUMN_NAME_COLUMN, cleanTask.getColumnName());
+                    row.putInt(TABLE_ID_COLUMN, cleanTask.getTableId());
+                    row.putInt(TABLE_TRUNCATE_VERSION, cleanTask.getTruncateVersion());
+                    row.putInt(COLUMN_TYPE_COLUMN, cleanTask.getColumnType());
+                    row.putInt(PARTITION_BY_COLUMN, cleanTask.getPartitionBy());
+                    row.putLong(UPDATED_TXN_COLUMN, cleanTask.getUpdatedTxn());
+                    row.putLong(COLUMN_VERSION_COLUMN,
+                            updatedColumnVersions.getQuick(i + ColumnVersionPurgeTask.OFFSET_COLUMN_VERSION));
+                    row.putTimestamp(PARTITION_TIMESTAMP_COLUMN,
+                            updatedColumnVersions.getQuick(i + ColumnVersionPurgeTask.OFFSET_PARTITION_TIMESTAMP));
+                    row.putLong(PARTITION_NAME_COLUMN,
+                            updatedColumnVersions.getQuick(i + ColumnVersionPurgeTask.OFFSET_PARTITION_NAME_TXN));
+                    row.append();
+                    updatedColumnVersions.setQuick(i + 3, Rows.toRowID(writer.getPartitionCount() - 1, writer.getTransientRowCount() - 1));
+                }
+            } catch (Throwable th) {
+                LOG.error().$("error saving to column version house keeping log, cannot append [table=").$(tableName).$(", error=").$(th).I$();
```

Comment:
we must tell used here how to housekeep columns manually, e.g. "please run vacuum...."

Reply:
done in [f660e8e](https://github.com/questdb/questdb/pull/1947/commits/f660e8e604ad2bb2b9c6fcf269bf44a2f3a4a750)